src/python/grpcio/grpc/_channel.py

```diff
@@ -263,35 +263,45 @@ def _rpc_state_string(class_name, rpc_state):
                 rpc_state.debug_error_string)
 
 
-class _RpcError(grpc.RpcError, grpc.Call):
+class _RpcError(grpc.RpcError, grpc.Call, grpc.Future):
     """An RPC error not tied to the execution of a particular RPC.
 
+    The state passed to _RpcError must be guaranteed not to be accessed by any
+    other threads.
+
+    The RPC represented by the state object must not be in-progress.
+
     Attributes:
       _state: An instance of _RPCState.
     """
 
     def __init__(self, state):
+        if state.cancelled:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for a cancelled RPC.")
+        if state.code is grpc.StatusCode.OK:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for a successfully completed RPC."
+            )
+        if state.code is None:
+            raise ValueError(
+                "Cannot instantiate an _RpcError for an incomplete RPC.")
```

Comment:
If we are copying state, maybe we can map above exception cases into corresponding `code` and `details`? E.g. if the `code` is `None` during initialization, maybe we can assign `UNKNOWN` to it.

Reply:
I don't see the advantages to this. Can you please elaborate?

Reply:
Since all tests are passing, I guess I'm talking about some corner cases here.

For the semantic of the `_RpcError` class, IMO it is responsible to construct an informative exception to propagate to application.

In cases that needs to raise an exception, we should prioritize the propagate of any debug details, and make sure the construction of the exception do not fail. If the initialization raises an exception, the original exception will be lost.

So, for cancelled RPC and incomplete RPC, I think it is fine to allow them to be `grpc.RpcError`. For successfully completed RPC, I agree we should raise an `ValueError` since that might indicate logic error in our framework.

Reply:
If these `ValueError`s were ever actually raised, it would be indicative of a bug in our own code. Now that I've seen the test suite pass with these checks in, I would rather just remove the checks completely.

Reply:
`ValueError`s removed.